frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,692 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The signed phase implementation.
+
+use super::*;
+use codec::{Encode, HasCompact};
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_npos_elections::{is_score_better, CompactSolution};
+use sp_runtime::Perbill;
+
+/// A raw, unchecked signed submission.
+///
+/// This is just a wrapper around [`RawSolution`] and some additional info.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, Default)]
+pub struct SignedSubmission<A, B: HasCompact, C> {
+	/// Who submitted this solution.
+	pub(crate) who: A,
+	/// The deposit reserved for storing this solution.
+	pub(crate) deposit: B,
+	/// The reward that should be given to this solution, if chosen the as the final one.
+	pub(crate) reward: B,
+	/// The raw solution itself.
+	pub(crate) solution: RawSolution<C>,
+}
+
+pub(crate) type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+pub(crate) type PositiveImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::PositiveImbalance;
+pub(crate) type NegativeImbalanceOf<T> = <<T as Config>::Currency as Currency<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+impl<T: Config> Pallet<T> {
+	/// Finish the singed phase. Process the signed submissions from best to worse until a valid one
+	/// is found, rewarding the best one and slashing the invalid ones along the way.
+	///
+	/// Returns true if we have a good solution in the signed phase.
+	///
+	/// This drains the [`SignedSubmissions`], potentially storing the best valid one in
+	/// [`QueuedSolution`].
+	pub fn finalize_signed_phase() -> (bool, Weight) {
+		let mut all_submission: Vec<SignedSubmission<_, _, _>> = <SignedSubmissions<T>>::take();
+		let mut found_solution = false;
+		let mut weight = T::DbWeight::get().reads(1);
+
+		while let Some(best) = all_submission.pop() {
+			let SignedSubmission { solution, who, deposit, reward } = best;
+			let active_voters = solution.compact.voter_count() as u32;
+			let feasibility_weight = {
+				// defensive only: at the end of signed phase, snapshot will exits.
+				let SolutionOrSnapshotSize { voters, targets } =
+					Self::snapshot_metadata().unwrap_or_default();
+				let desired_targets = Self::desired_targets().unwrap_or_default();
+				let v = voters;
+				let t = targets;
+				let a = active_voters;
+				let w = desired_targets;
+				T::WeightInfo::feasibility_check(v, t, a, w)
```

Comment:
```suggestion
				T::WeightInfo::feasibility_check(
					voters, 
					targets, 
					active_voters, 
					desired_targets,
				)
```

Less cognitive load.